src/pkg/packager/composer/list.go

```diff
@@ -0,0 +1,420 @@
+	} else {
+		tb = filepath.Join(cache, "blobs", "sha256", componentDesc.Digest.Encoded())
+		dir = filepath.Join(cache, "dirs", componentDesc.Digest.Encoded())
+
+		store, err := ocistore.New(cache)
+		if err != nil {
+			return err
+		}
+
+		ctx := context.TODO()
+		// ensure the tarball is in the cache
+		exists, err := store.Exists(ctx, componentDesc)
+		if err != nil {
+			return err
+		} else if !exists {
+			copyOpts := remote.CopyOpts
+			// TODO: investigate why the default FindSuccessors in CopyWithProgress is not working
+			copyOpts.FindSuccessors = content.Successors
+			if err := remote.CopyWithProgress([]ocispec.Descriptor{componentDesc}, store, copyOpts, cache); err != nil {
+				return err
+			}
+			exists, err := store.Exists(ctx, componentDesc)
+			if err != nil {
+				return err
+			} else if !exists {
+				return fmt.Errorf("failed to fetch remote component: %+v", componentDesc)
+			}
+		}
+	}
+
+	if err := utils.CreateDirectory(dir, 0700); err != nil {
+		return err
+	}
+
+	cwd, err := os.Getwd()
+	if err != nil {
+		return err
+	}
+	rel, err := filepath.Rel(cwd, dir)
+	if err != nil {
+		return err
+	}
+	// the tail node is the only node whose relativeToHead is based solely upon cwd<->cache
+	// contrary to the other nodes, which are based upon the previous node
+	ic.tail.relativeToHead = rel
+
+	if oci.IsEmptyDescriptor(componentDesc) {
+		// nothing was fetched, nothing to extract
+		return nil
+	}
+
+	tu := archiver.Tar{
+		OverwriteExisting: true,
+		// removes /<name>/ from the paths
+		StripComponents: 1,
+	}
+	return tu.Unarchive(tb, dir)
+}
+
+// Compose merges the import chain into a single component
+// fixing paths, overriding metadata, etc
+func (ic *ImportChain) Compose() (composed types.ZarfComponent, err error) {
+	composed = ic.tail.ZarfComponent
+
+	if ic.tail.prev == nil {
+		// only had one component in the import chain
+		return composed, nil
+	}
+
+	if err := ic.fetchOCISkeleton(); err != nil {
+		return composed, err
+	}
+
+	// fix paths on the tail
+	fixPaths(&ic.tail.ZarfComponent, ic.tail.relativeToHead)
+	// save mutated tail
+	composed = ic.tail.ZarfComponent
+
+	// dont compose 2x
+	node := ic.tail.prev
+	for node != nil {
+		fixPaths(&node.ZarfComponent, node.relativeToHead)
+
+		// perform overrides here
+		overrideMetadata(&composed, node.ZarfComponent)
+		overrideDeprecated(&composed, node.ZarfComponent)
+		overrideResources(&composed, node.ZarfComponent)
+		overrideActions(&composed, node.ZarfComponent)
+
+		composeExtensions(&composed, node.ZarfComponent, node.relativeToHead)
+
+		node = node.prev
+	}
+
+	return composed, nil
+}
+
+// MergeVariables merges variables from the import chain
+func (ic *ImportChain) MergeVariables(existing []types.ZarfPackageVariable) (merged []types.ZarfPackageVariable) {
+	exists := func(v1 types.ZarfPackageVariable, v2 types.ZarfPackageVariable) bool {
+		return v1.Name == v2.Name
+	}
+
+	merged = helpers.MergeSlices(existing, merged, exists)
```

Comment:
It may be better to initialize the import chain with the variables / constants instead of having to handle the cases here.

Reply:
The reason for this is because it maintains an import chains "knowledge" of variables and consts. Since `existing` is going to mutate from chain to chain, saving it into the chain means the chain has knowledge of things outside itself.